### PR TITLE
feat: added --follow option to `hyprctl rollinglog` #6300

### PR DESCRIPTION
**Pull request by Truenya**
#### Describe your PR, what does it fix/add?
Added --follow option to `hyprctl rollinglog` 
#6300 

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
Also added `onlybuild` command to Makefile (it is annoying to completely rebuild whole project any time).

#### Is it ready for merging, or does it need work?
Ready, working, tested.

### CODE REVIEW

**Comment by vaxerski** [comment]
tbh I fail to see the use - one can always like... `watch`?

**Comment by zjeffer** [comment]
Using watch is kind of annoying and very limiting: `hyprctl rollinglog` only shows you the last 50 lines, so using watch (with for example `-n 0.1` to make sure you get every line) will keep showing the last 50 lines meaning you can't scroll up. I sometimes want to know what happened in the last 100 lines, for example, so I need the ability to scroll up in the logs. 

I'll test this PR tomorrow, but from the description it does exactly what I wanted.

**Comment by Truenya** [comment]
https://www.youtube.com/watch?v=JIz6l6Xlouc

**Comment by rtgiskard** [comment]
As rollinglog can not follow the output like `jounalctl -f` does, I'm generally using `tail -f /run/user/1000/hypr/*/hyprland.log` instead.

It would be good if rollinglog can follow the log, but I dont't really like current implemention with keep sending log in a thread via the socket, maybe something leke a wrapper for tail is simple and efficient enough.

And, perhaps, use something like spdlog, and/or with syslog as backend, then just use journalctl to manage the log.

**Comment by Truenya** [comment]
>  I'd prefer a wrapper for tail which would be simple and efficient enough

Small detail: rollinglog works independently from any other logging methods. So we cannot use tail wrapper here. If the user does not want to write the log to a file, we cannot force him.

**Comment by zjeffer** [comment]
> rollinglog works independently from any other logging methods

What do you mean with this? 

When I initially created the issue, my request was basically to make `hyprctl rollinglog -f` be a wrapper around `tail -f <logfile>`, without having to specify the whole path to the logfile.

**Comment by Truenya** [comment]
@zjeffer You can disable logging in the configuration and still get it with the rollinglog command. The added --follow option preserves this logic.

File: Makefile [hunk 1]
```diff
@@ -105,3 +105,6 @@ asan:
 	@echo "Hyprland done"
 
 	ASAN_OPTIONS="detect_odr_violation=0,log_path=asan.log" HYPRLAND_NO_CRASHREPORTER=1 ./build/Hyprland -c ~/.config/hypr/hyprland.conf
+
+onlybuild:
```
Comment (vaxerski) [inline Makefile:109]: I think this is unintentional?

File: hyprctl/main.cpp [hunk 2]
```diff
@@ -299,6 +347,9 @@ int main(int argc, char** argv) {
                 fullArgs += "a";
             } else if ((ARGS[i] == "-c" || ARGS[i] == "--config") && !fullArgs.contains("c")) {
                 fullArgs += "c";
+            } else if ((ARGS[i] == "-f" || ARGS[i] == "--follow") && !fullArgs.contains("f")) {
```
Comment (vaxerski) [inline hyprctl/main.cpp:350]: would be nice to add to the help menu

**Review by vaxerski — COMMENTED** [review]
additionally:
 - clang-format needed
 - no {} around short ifs (too many cases to label all)
 - wiki mr needed

**Comment by Truenya** [comment]
> * wiki mr needed

can u please explain more concrete? what exactly needs to be described?

**Comment by Truenya** [comment]
> no {} around short ifs (too many cases to label all)

do you mean that it is necessary to remove {} for short Ifs? as far as I can see they are everywhere in MR

File: Makefile (hunk 1 above)
Comment (Truenya) [reply Makefile]: Done

File: hyprctl/main.cpp (hunk 2 above)
Comment (Truenya) [reply hyprctl/main.cpp:349]: Done

**Comment by vaxerski** [comment]
the wiki mr yea thats all, just to let people know

for ifs: 

we always
```cpp
if (a) 
   b();
```
in hyprland

**Comment by Truenya** [comment]
> the wiki mr yea thats all, just to let people know
> 
> for ifs:
> 
> we always
> 
> ```c++
> if (a) 
>    b();
> ```
> 
> in hyprland

Done

File: src/debug/RollingLogFollow.hpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+        void StopFor(int socket) {
+            std::unique_lock<std::shared_mutex> w(m);
+            socketToRollingLogFollowQueue.erase(socket);
+            if (socketToRollingLogFollowQueue.empty())
+                running = false;
+        }
+        void StartFor(int socket) {
+            std::unique_lock<std::shared_mutex> w(m);
+            socketToRollingLogFollowQueue[socket] = "[LOG] Following log to socket: " + std::to_string(socket) + " started" + "\n";
```
Comment (vaxerski) [inline src/debug/RollingLogFollow.hpp:53]: std::format

File: src/debug/RollingLogFollow.hpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        // Returns true if the queue is empty for the given socket
+        bool isEmpty(int socket) {
+            std::shared_lock<std::shared_mutex> r(m);
+            return socketToRollingLogFollowQueue[socket].empty();
+        }
+
+        std::string DebugInfo() {
+            std::shared_lock<std::shared_mutex> r(m);
+            return "RollingLogFollow, got " + std::to_string(socketToRollingLogFollowQueue.size()) + " connections";
```
Comment (vaxerski) [inline src/debug/RollingLogFollow.hpp:20]: std::format

File: src/debug/HyprCtl.cpp [hunk 5]
```diff
@@ -1727,6 +1728,45 @@ std::string CHyprCtl::makeDynamicCall(const std::string& input) {
...
+
+            usleep(100000);
+        }
+        close(conn);
+        Debug::RollingLogFollow::Get().StopFor(conn);
+    }).detach();
+}
+
+bool isFollowUpRollingLogRequest(const std::string& request) {
+    return request.contains("rollinglog") && request.find_first_of('f') != std::string::npos;
```
Comment (vaxerski) [inline src/debug/HyprCtl.cpp:1767]: .contains("f")

File: src/debug/HyprCtl.cpp [hunk 6]
```diff
@@ -1727,6 +1728,45 @@ std::string CHyprCtl::makeDynamicCall(const std::string& input) {
...
+    std::thread([conn]() {
+        while (Debug::RollingLogFollow::Get().IsRunning()) {
+            if (Debug::RollingLogFollow::Get().isEmpty(conn)) {
+                sleep(1);
+                continue;
+            }
+
+            auto line = Debug::RollingLogFollow::Get().GetLog(conn);
+            if (!successWrite(conn, line))
+                // We cannot write, when connection is closed. So thread will successfully exit by himself
```
Comment (vaxerski) [inline src/debug/HyprCtl.cpp:1756]: by itself

File: hyprctl/main.cpp [hunk 7]
```diff
@@ -90,13 +91,53 @@ std::vector<SInstanceData> instances() {
...
+    sigintReceived = true;
+    std::cout << "[hyprctl] SIGINT received, closing connection" << std::endl;
+}
+
+int rollingRead(const int socket) {
+    sigintReceived = false;
+    signal(SIGINT, intHandler);
+
+    constexpr size_t BUFFER_SIZE         = 8192;
+    char             buffer[BUFFER_SIZE] = {0};
```
Comment (vaxerski) [inline hyprctl/main.cpp:105]: use an std::array please

File: hyprctl/main.cpp [hunk 8]
```diff
@@ -90,13 +91,53 @@ std::vector<SInstanceData> instances() {
...
+            close(socket);
+            return 5;
+        }
+
+        if (sizeWritten == 0)
+            break;
+
+        if (sizeWritten > 0) {
+            std::cout << std::string(buffer, sizeWritten);
+            memset(buffer, 0, BUFFER_SIZE);
```
Comment (vaxerski) [inline hyprctl/main.cpp:122]: when you use std::array, instead of unsafe memset, you can `.fill()`

File: src/debug/HyprCtl.cpp [hunk 9]
```diff
@@ -1727,6 +1728,45 @@ std::string CHyprCtl::makeDynamicCall(const std::string& input) {
...
+    return false;
+}
+
+void runWritingDebugLogThread(const int conn) {
+    Debug::log(LOG, "In followlog thread, got connection, start writing: {}", conn);
+    //will be finished, when reading side close connection
+    std::thread([conn]() {
+        while (Debug::RollingLogFollow::Get().IsRunning()) {
+            if (Debug::RollingLogFollow::Get().isEmpty(conn)) {
+                sleep(1);
```
[the remark this answers is not in the document]
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: done

File: src/debug/HyprCtl.cpp [hunk 10]
```diff
@@ -1727,6 +1728,45 @@ std::string CHyprCtl::makeDynamicCall(const std::string& input) {
...
+                continue;
+            }
+
+            auto line = Debug::RollingLogFollow::Get().GetLog(conn);
+            if (!successWrite(conn, line))
+                // We cannot write, when connection is closed. So thread will successfully exit by himself
+                break;
+
+            usleep(100000);
```
[the remark this answers is not in the document]
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: done

File: src/debug/HyprCtl.cpp [hunk 11]
```diff
@@ -1770,9 +1810,16 @@ int hyprCtlFDTick(int fd, uint32_t mask, void* data) {
...
-    write(ACCEPTEDCONNECTION, reply.c_str(), reply.length());
+    successWrite(ACCEPTEDCONNECTION, reply);
 
-    close(ACCEPTEDCONNECTION);
+    if (isFollowUpRollingLogRequest(request)) {
+        Debug::log(LOG, "Followup rollinglog request received. Starting thread to write to socket.");
+        Debug::RollingLogFollow::Get().StartFor(ACCEPTEDCONNECTION);
+        runWritingDebugLogThread(ACCEPTEDCONNECTION);
+        Debug::log(LOG, Debug::RollingLogFollow::Get().DebugInfo());
+    } else {
```
[the remark this answers is not in the document]
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: done

File: src/debug/RollingLogFollow.hpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            std::shared_lock<std::shared_mutex> r(m);
+            return running;
+        }
+
+        void StopFor(int socket) {
+            std::unique_lock<std::shared_mutex> w(m);
+            socketToRollingLogFollowQueue.erase(socket);
+            if (socketToRollingLogFollowQueue.empty())
+                running = false;
+        }
```
[the remark this answers is not in the document]
Comment (Truenya) [reply src/debug/RollingLogFollow.hpp:50]: done

File: src/debug/RollingLogFollow.hpp (hunk 3 above)
Comment (Truenya) [reply src/debug/RollingLogFollow.hpp]: done

File: src/debug/RollingLogFollow.hpp (hunk 4 above)
Comment (Truenya) [reply src/debug/RollingLogFollow.hpp]: done

File: src/debug/HyprCtl.cpp (hunk 5 above)
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: done

File: src/debug/HyprCtl.cpp (hunk 6 above)
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: done

File: hyprctl/main.cpp (hunk 7 above)
Comment (Truenya) [reply hyprctl/main.cpp]: done

File: hyprctl/main.cpp (hunk 8 above)
Comment (Truenya) [reply hyprctl/main.cpp]: done

**Comment by vaxerski** [comment]
this MR will clash with #6380 and since that one is smaller can you please wait for until it's merged and then just fixup this one (use `log` instead of std::cout for ignorable logs)

**Comment by vaxerski** [comment]
6380 is merged, can you please update the MR?

**Comment by Truenya** [comment]
@vaxerski done

File: hyprctl/main.cpp [hunk 13]
```diff
@@ -366,6 +415,11 @@ int main(int argc, char** argv) {
         return 0;
     }
 
+    if (needRoll && !fullRequest.contains("/rollinglog")) {
+        std::cout << "only 'rollinglog' command supports '--follow' option" << std::endl;
```
Comment (vaxerski) [inline hyprctl/main.cpp:419]: should be `log()`

File: src/debug/HyprCtl.cpp [hunk 14]
```diff
@@ -1723,6 +1724,48 @@ std::string CHyprCtl::makeDynamicCall(const std::string& input) {
...
+
+void runWritingDebugLogThread(const int conn) {
+    using namespace std::chrono_literals;
+    Debug::log(LOG, "In followlog thread, got connection, start writing: {}", conn);
+    //will be finished, when reading side close connection
+    std::thread([conn]() {
+        while (Debug::RollingLogFollow::Get().IsRunning()) {
+            if (Debug::RollingLogFollow::Get().isEmpty(conn)) {
+                std::this_thread::sleep_for(1000ms);
+//                sleep(1);
```
Comment (vaxerski) [inline src/debug/HyprCtl.cpp:1748]: why leave these?

File: hyprctl/main.cpp [hunk 15]
```diff
@@ -98,13 +99,54 @@ std::vector<SInstanceData> instances() {
...
+    sigintReceived = true;
+    std::cout << "[hyprctl] SIGINT received, closing connection" << std::endl;
+}
+
+int rollingRead(const int socket) {
+    sigintReceived = false;
+    signal(SIGINT, intHandler);
+
+    constexpr size_t BUFFER_SIZE         = 8192;
+//    char             buffer[BUFFER_SIZE] = {0};
```
Comment (vaxerski) [inline hyprctl/main.cpp:113]: why is this left

File: hyprctl/main.cpp (hunk 13 above)
Comment (Truenya) [reply hyprctl/main.cpp]: fixed

File: src/debug/HyprCtl.cpp (hunk 14 above)
Comment (Truenya) [reply src/debug/HyprCtl.cpp]: fixed

File: hyprctl/main.cpp (hunk 15 above)
Comment (Truenya) [reply hyprctl/main.cpp:113]: fixed

Comment (vaxerski) [reply hyprctl/main.cpp:113]: is it though

Comment (Truenya) [reply hyprctl/main.cpp]: oh sorry
now really done

**Comment by vaxerski** [comment]
conflict

**Comment by Truenya** [comment]
@vaxerski resolved

**Review by vaxerski — APPROVED** [review]
yabba dabba doo
